index.html

```diff
@@ -3930,6 +3938,64 @@ <h2>Identification</h2>
+  removes variations between different expressions of the same information model,
+  so that the same information always results in the same JSON serialization.
+  This unique representation is called a <a>Canonical TD</a>.
+  Canonical TDs are important for applications that need to compare TDs as strings,
+  for example, to compute signatures on TDs or to compute differences between TDs.
+  Whenever there is a choice of multiple ways to present information, canonicalization
+  selects only one. There are many places in which optional features are included in JSON,
+  such as white space, or where different representation formats are possible, such as
+  the number of digits in a number representation or the order of elements in an object.
+  </p><p>
+  Some of the choices leading to the efficient computation of a canonical form, 
+  such as the suppression of white space and sorting of elements in objects 
+  do make the representation less human readable.
+  In general, the canonical form presented here is optimized for machine-to-machine
+  use cases such as signing and to accommodate devices with limited network and
+  compute capabilities.  Compactness and ease of implementation
+  has been emphasized over human readability.
+  </p>
+  <p>Computing a canonical form directly from an RDF triple representation is
+  difficult so the starting point is the JSON serialization of a TD.  
+  The following sequence of transformations are then applied to this serialization: 
+  <ol>
+  <li>Remove any elements that are given their default value. 
+  For example, the default value of <code>observable</code>
+  is <code>false</code>.  
+  If an input TD explictly assigns the value <code>false</code> to <code>observable</code>,
+  this element and the redundant value will be removed.</li>
+  <li>Rewrite values that can be expressed as either an array or as a single value 
+  as a single value if there is only one element.  
+  In other words, remove square brackets around
+  arrays of single elements where this is permitted.
+  </li>
+  <li>For all extension vocabulary terms, sort the base URIs, assign system-generated prefixes 
+  using the template <code>e{n}</code> (where <code>{n}</code> is an integer sequence number starting
+  at 1), and use these prefixes in place of base URIs. Note: some systems discard the prefixes
+  and expand all vocabulary terms into URIs.  This rule allows the canonical form to be
+  generated even in this case. It does, however, lead to the loss of the original prefixes.</li>
+  <li>Apply the JSON Canonicalization Scheme [[!RFC8785]], which among other transformations,
+  sorts object members by name, removes white space, and normalizes number representations.</li>
+  <li>As an exception to the sorted order,
+  move <code>securityDefinitions</code> to just before <code>security</code> at the top level.
+  This is to make single-pass processing easier, so these definitions 
+  occur prior to their first use.</li> 
```

Comment:
The second bullet is going to be very hard to implement. Basically, we will not be able to use existing processors which implement the likes of RFC8785 JSON Canonicalization Scheme.

I suggest leaving consumer requirements out of this. Otherwise, there will be many more reordering exceptions: e.g. `base` before `actions` and more when we dig into each object tree.

Reply:
Good point about base.  We either need to do this for all such things, or not.

Reply:
Anyhow, as stated this is broken, because "security" also needs to come before "action".  I think what I'll do is just specify the order the top-level members of Thing need to be listed in.  We can, however, separately decide if we even want to require an ordering different from just a plain sort.